pkg/controllers/elasticsearch/actions/scale.go

```diff
@@ -0,0 +1,212 @@
+	// The node pool being scaled
+	NodePool *v1alpha1.ElasticsearchClusterNodePool
+	// Number of replicas to scale to
+	Replicas int32
+}
+
+var _ controllers.Action = &Scale{}
+
+func (c *Scale) Name() string {
+	return "Scale"
+}
+
+func (c *Scale) Message() string {
+	return fmt.Sprintf("Scaled node pool %q to %d replicas", c.NodePool.Name, c.NodePool.Replicas)
+}
+
+// Execute will scale the s.NodePool statefulset to the desired number of
+// replicas. It will refuse to scale if the cluster is not in a 'state to scale'
+// as defined by s.canScaleNodePool.
+func (s *Scale) Execute(state *controllers.State) error {
+	if s.NodePool == nil || s.Cluster == nil {
+		return fmt.Errorf("internal error: nodepool and cluster must be set")
+	}
+
+	statefulSetName := util.NodePoolResourceName(s.Cluster, s.NodePool)
+	statefulSet, err := state.StatefulSetLister.StatefulSets(s.Cluster.Namespace).Get(statefulSetName)
+	if k8sErrors.IsNotFound(err) {
+		state.Recorder.Eventf(s.Cluster, core.EventTypeWarning, "Err"+s.Name(), "Cannot get statefulset %q for node pool %q: %v", statefulSetName, s.NodePool.Name, err)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	currentReplicas := statefulSet.Spec.Replicas
+	// TODO: not sure if we should treat nil as 1 or 0 instead of erroring
+	if currentReplicas == nil {
+		return fmt.Errorf("current number of replicas on statefulset cannot be nil")
+	}
+	replicaDiff := s.Replicas - (*currentReplicas)
+	// if this scale won't change the number of replicas, we can return early
+	if replicaDiff == 0 {
+		return nil
+	}
+
+	canScale, err := s.canScaleNodePool(state, statefulSet, replicaDiff)
+	if err != nil {
+		return err
+	}
+	if !canScale {
+		return nil
+	}
+
+	ssCopy := statefulSet.DeepCopy()
+	ssCopy.Spec.Replicas = &s.Replicas
+	_, err = state.Clientset.AppsV1beta1().StatefulSets(ssCopy.Namespace).Update(ssCopy)
+	if err != nil {
+		return err
+	}
+	state.Recorder.Eventf(s.Cluster, core.EventTypeNormal, s.Name(), "Scaled node pool %q to %d replicas", s.NodePool.Name, s.NodePool.Replicas)
+
+	return nil
+}
+
+// canScaleNodePool will determine whether this scale operation is 'valid'.
+// Valid is defined as:
+// - if the node pool does not have a 'data' node pool role, a scale is always
+//   valid.
+// - if the scale is not a 'scale down', it is always valid.
+// - if all pilots effected by the scale (e.g. those that would be removed)
+//   have been drained of all shards, we can scale down.
+// - otherwise reject the scale down.
+func (s *Scale) canScaleNodePool(state *controllers.State, statefulSet *apps.StatefulSet, replicaDiff int32) (bool, error) {
+	// always allow scale up, or staying the same (no-op)
+	if replicaDiff >= 0 {
+		return true, nil
+	}
+	// we can always scale down non-data roles, as validation should
+	// ensure that at least one data, ingest and master node exists already
+	if !utilapi.ContainsElasticsearchRole(s.NodePool.Roles, v1alpha1.ElasticsearchRoleData) {
+		return true, nil
+	}
```

Comment:
Perhaps - but this could create unstable clusters, and potentially then causing a lot of churn in control loops (Pilots might be updating their resources with error conditions frequently, causing the navigator control loop to keep resyncing).

I'm not against scale-to-zero, but I don't think it needs to be incorporated right now/as part of this PR?